File: 1695-maximum-erasure-value/1695-maximum-erasure-value.py
class Solution:
    def maximumUniqueSubarray(self, nums: List[int]) -> int:
        n = len(nums)
        st = set()
        ans = -1
        sm = 0 # holds sum of unique numberes subarray
        i = 0
        j = 0
        while j < n:
            if nums[j] not in st:
                sm += nums[j]
                st.add(nums[j])
            else:
                ans = max(ans, sm)
                while nums[j] in st:
                    sm -= nums[i]
                    st.discard(nums[i])
                    i += 1
                sm += nums[j]
                st.add(nums[j])
            j += 1
        return max(ans, sm)

File: 1695-maximum-erasure-value/1695-maximum-erasure-value.cpp
class Solution {
public:
    int maximumUniqueSubarray(vector<int>& nums) {
        int n = nums.size();
        unordered_set<int> st;
        int ans = INT_MIN;
        int sum = 0;
        int i=0, j=0;
        while(j < n){
            if(!st.count(nums[j])){
                sum += nums[j];
                st.insert(nums[j]);
                j++;
            }
            else{
                ans = max(ans, sum);
                while(st.count(nums[j])){
                    sum -= nums[i];
                    st.erase(nums[i]);
                    i++;
                }
                st.insert(nums[j]);
                sum += nums[j];
                j++;
            }
        }
        return max(ans, sum); 
    }
};